File: go/pkg/grovex/a_render.go
package grovex

import (
	"github.com/kmirzavaziri/grove/go/pkg/gex"
	"github.com/kmirzavaziri/grove/go/pkg/grove"
)

type ARenderPayload struct {
	NodePath      []string
	Request       *gex.Value
	Node          *grove.RenderedNode
	UpdateHistory bool
	Patch         bool
}

func ARender(payload ARenderPayload) *grove.Action {
	return &grove.Action{
		Type:    "grovex.ARender",
		Payload: gex.Marshal(payload),
	}
}

File: ts/packages/grovex/src/ARender.ts
import {AppContextValue, ComponentProps, modifyComponentProps, reverseRoute} from '@grove/grove';
import type {Struct} from '@grove/grove/src/value';

export interface ARenderProps {
    node_path: string[];
    request?: Struct;
    node?: ComponentProps;
    patch: boolean;
    update_history?: boolean;
}

export function ARender(appContextValue: AppContextValue, props: ARenderProps): void {
    if (props.update_history) {
        window.history.pushState({}, '', reverseRoute(props.node_path, props.request || {}));
    }

    const modify = (newNode: ComponentProps) => (node: ComponentProps) => {
        modifyComponentProps(node, newNode, props.patch);
    };

    if (props.node) {
        return appContextValue.dispatch({path: props.node_path, node: modify(props.node)});
    }

    appContextValue.apiHandlers
        .fetch(props.node_path, props.request)
        .then((node) => appContextValue.dispatch({path: props.node_path, node: modify(node)}));
}
